File: spiral/src/spiral.js
;(function (root) {

  // Constants
  var HEIGHT = 500;
  var WIDTH = 500;
  var BACKGROUND = '#ddd';
  var LINE_COLOR = '#333';
  var LINE_WIDTH = 2;

  // References
  var canvas = document.getElementById('spiral');
  var ctx = canvas.getContext('2d');

  // Math aliases
  var E = Math.E;
  var PI = Math.PI;
  var abs = Math.abs;
  var atan2 = Math.atan2;
  var cos = Math.cos;
  var pow = Math.pow;
  var sin = Math.sin;
  var sqrt = Math.sqrt;

  // Helpers
  function range (size) {
    var i, arr = [];

    for (i = 0; i < size; i++) {
      arr.push(i);
    }

    return arr;
  }

  // Drawing functions
  function spiral (
    startX, // number
    startY, // number
    endX, // number
    endY, // number
    decay, // number
    length, // number
    draw // boolean
  ) {
    draw = draw || true;

    if (-1e-2 < decay && decay < 1e-2) {
      ctx.moveTo(startX, startY);
      ctx.lineTo(endX, endY);

      return;
    }

    var b = 1 / decay;
    var dt = PI / 4;
    var m = 8;

    if (abs(b) > 1) {
      dt /= abs(b);
    }

    if (b < 0) {
      dt *= -1;
    }

    var D = sqrt(pow(startX - endX, 2), pow(startY - endY, 2));
    var t = atan2(startY - endY, startX - endX);
    var a = pow(E, -b * t) * D;
    ctx.moveTo(a * pow(E, b * t) * cos(t) + endX,
        a * pow(E, b * t) * sin(t) + endY);

    range(length * m).forEach(function () {
      var t2 = t - dt;
      var Dt = t2 - t;

      // Logarithmic spiral equations
      var x0 = pow(a * E, b * t) * cos(t) + endX;
      var y0 = pow(a * E, b * t) * sin(t) + endY;

      var x3 = pow(a * E, b * t2) * cos(t2) + endX;
      var y3 = pow(a * E, b * t2) * sin(t2) + endY;

      // Derivatives of spiral equation
      var dx1 = pow(a * b * E, b * t) * cos(t) - pow(a * E, b * t) * sin(t);
      var dy1 = pow(a * E, b * t) * cos(t) + pow(a * b * E, b * t) * sin(t);

      var dx2 = pow(a * b * E, b * t2) * cos(t2) - pow(a * E, b * t2) * sin(t2);
      var dy2 = pow(a * E, b * t2) * cos(t2) + pow(a * b * E, b * t2) * sin(t2);

      // Calculate control points
      var x1 = x0 + ((Dt / 3) * dx1);
      var y1 = y0 + ((Dt / 3) * dy1);

      var x2 = x3 + ((Dt / 3) * dx2);
      var y2 = y3 + ((Dt / 3) * dy2);
      t -= dt;

      ctx.bezierCurveTo(x1, y1, x2, y2, x3, y3);
    });
  }

  // Setup
  canvas.height = HEIGHT;
  canvas.style.height = HEIGHT + 'px';
  canvas.width = WIDTH;
  canvas.style.width = WIDTH + 'px';
  canvas.style.background = BACKGROUND;

  // Animation loop
  (function update() {
    webkitRequestAnimationFrame(update);

    // Clear the canvas
    canvas.width = canvas.width;

    var beginX = 30;
    var beginY = 200;
    var endX = 300;
    var endY = 200;
    var decay = 5;
    var rotations = 2;

    // Start drawing operations
    ctx.beginPath();
    ctx.lineWidth = LINE_WIDTH;

    spiral(beginX, beginY, endX, endY, decay, rotations);

    // End drawing operations
    ctx.strokeStyle = LINE_COLOR;
    ctx.stroke();
    ctx.closePath();
  }());

} (this));

File: spiral/reference/spiral.py
# Taken from http://nodebox.net/code/index.php/shared_2008-02-26-23-26-39
# Pasting this here because I want to learn from it.
# All credit goes to "Mark M." (See source link)
from nodebox.graphics import BezierPath

class spiral(BezierPath):
    '''A logarithmic spiral beginning at start_x, start_y and spiraling toward the point end_x, end_y. 
       Decay: determines how tightly the spiral turns. Negative values for decay result in the spiral turning
       opposite direction. A value of zero results in a line
       Length: is how many times around the spiril turns.
    '''
    def __init__(self, start_x, start_y, end_x, end_y, decay, length, draw=True, **kwargs):
        from math import sin, cos, pi, e, sqrt, atan2, pow
       
        BezierPath.__init__(self, _ctx,  **kwargs) 
        if  -1e-2 < decay < 1e-2: 
            # very decay of 0 indicates a line
            # but very small decays are practically line ans too small will cause e**1/decay to big = python error
            self.moveto(start_x, start_y )
            self.lineto(end_x, end_y )
            self.inheritFromContext(kwargs.keys())
            if draw:
                self.draw()
            return
        b = 1/float(decay)
        dt = pi/4
        m = 8
        if abs(b) > 1:
            # when the curve gets very flat we need to take a few more samples
            # a better approach might be to simply fit a single cubic curve to the end points
            dt /= abs(b)
        if b < 0:
            #curve reverses    
            dt *= -1
        D = sqrt(pow(start_x-end_x,2) + pow(start_y-end_y,2))
        t =  atan2(start_y-end_y, start_x-end_x)
        a = e**(-b*t)*D
        self.moveto(a*e**(b*t) * cos(t) + end_x, a*e**(b*t) * sin(t) + end_y )

        for i in range(int(length*m)):
            t2 = t - dt
            Dt = t2 -t
            # logarithmic spiral equations 
            x0 = a*e**(b*t) * cos(t) + end_x
            y0 = a*e**(b*t) * sin(t) + end_y 
        
            x3 = a*e**(b*t2) * cos(t2) + end_x
            y3 = a*e**(b*t2) * sin(t2) + end_y
  
            # Derivatives of spiral equation
            dx1 = a*b*e**(b*t) * cos(t) - a*e**(b*t) * sin(t)
            dy1 = a*e**(b*t) * cos(t) + a*b*e**(b*t) * sin(t) 

            dx2 = a*b*e**(b*t2) * cos(t2) - a*e**(b*t2) * sin(t2)
            dy2 = a*e**(b*t2) * cos(t2) + a*b*e**(b*t2) * sin(t2)  
      
            # calculate control points
            x1 = x0 + ((Dt/3.0) * dx1)
            y1 = y0 + ((Dt/3.0) * dy1)
   
            x2 = x3 - ((Dt/3.0) * dx2)
            y2 = y3 - ((Dt/3.0) * dy2)
            t -= dt
            self.curveto(x1, y1, x2, y2, x3, y3)
        self.inheritFromContext(kwargs.keys())
        if draw:
           self.draw()

#### End Spiral Class

size(820, 820)
autoclosepath(close=False)
nofill()
strokewidth(.5)
var("begin_x", NUMBER, 30, 0, WIDTH)
var("begin_y", NUMBER, 200, 0, WIDTH)
var("end_x", NUMBER, 300, 0, WIDTH)
var("end_y", NUMBER, 200, 0, WIDTH)
var("decay", NUMBER, 5, -10, 10)
var("rotations", NUMBER, 2, .25, 5)
p = spiral(begin_x, begin_y, end_x, end_y, decay , rotations, stroke=(.7, .2, .2))